lib/Bindings/Python/circt/dialects/esi.py

```diff
@@ -4,3 +4,318 @@
+            (inst["outer_sym"], inst["inner"]) for inst in svc["instance_path"]
+        ]:
+          m = self._get_module(inner_ref[0])
+          parent.instances[inner_ref[1]] = m
+          parent = m
+
+    # For any modules which have services, add them as appropriate.
+    for mod in self.services["modules"]:
+      m = self._get_module(mod["symbol"])
+      for svc in mod["services"]:
+        m.services.append(svc)
+
+  def _get_module(self, mod_sym: str):
+    """Get a module adding an entry if it doesn't exist."""
+    if mod_sym not in self.modules:
+      self.modules[mod_sym] = SoftwareApiBuilder.Module(mod_sym)
+    return self.modules[mod_sym]
+
+  def build(self, os: FileIO):
+    """Output the API (in a pre-determined order) via callbacks. Encourages some
+    level of consistency between language APIs."""
+
+    self.os = os
+    self._write_header()
+
+    for decl in self.services["declarations"]:
+      self._write_decl(decl)
+    os.write("\n\n")
+
+    self._write_namespace("modules")
+    for mod in self.modules.values():
+      self._write_module(mod)
+    os.write("\n\n")
+
+    # Need to write the modules first since it populates the types set as a side
+    # effect.
+    self._write_namespace("types")
+    for type_name, type_dict in self.types.items():
+      self._write_type(type_name, type_dict)
+
+    self._writeline()
+    self._writeline()
+    for top in self.services["top_levels"]:
+      self._write_top(self._get_module(top["module"][1:]))
+    self.os = None
+
+  def _writeline(self, line: str = ""):
+    self.os.write(line + "\n")
+
+  def _write_header(self):
+    pass
+
+  def _write_namespace(self, namespace: str):
+    pass
+
+  def _write_decl(self, decl: Dict):
+    assert False, "Unimplemented"
+
+  def _write_module(self, mod):
+    # mod is SoftwareApiBuilder.Module.
+    assert False, "Unimplemented"
+
+  def _write_type(self, name: str, type: Dict):
+    assert False, "Unimplemented"
+
+  def _write_top(self, top: Dict):
+    assert False, "Unimplemented"
+
+  def get_type_name(self, type: Dict):
+    """Create a name for 'type', record it, and return it."""
+    if "capnp_name" in type:
+      name = type["capnp_name"]
+    else:
+      name = "".join([c if c.isalnum() else '_' for c in type["mlir_name"]])
+    self.types[name] = type
+    return name
+
+
+class PythonApiBuilder(SoftwareApiBuilder):
+
+  def __init__(self, services_json: str, capnp_schema: Optional[str]):
+    super().__init__(services_json, capnp_schema)
+
+  def build(self, system_name: str, output_dir: pathlib.Path):
+    """Emit a Python ESI runtime library into 'output_dir'."""
+    libdir = output_dir / "esi_rt"
+    if not libdir.exists():
+      libdir.mkdir()
+
+    # Create __init__.py and copy in the standard files.
+    init_file = libdir / "__init__.py"
+    init_file.touch()
+    common_file = libdir / "common.py"
+    common_file.write_text((__dir__ / "esi_runtime_common.py").read_text())
```

Comment:
What is the reasoning behind copying a common file instead of including libraries distributed within the PyCDE package, seeing as that common file isn't parameterized? That would seem more natural and maintainable to me.

e.g.
- for python "pycde/esi/api/py/common.py"
- for c++ "pycde/esi/api/cpp/common.h, (cmake includes, ...), ..."

Reply:
These are runtimes which are meant to be distributed with the compiled image, not just used with the compiler.